use tsv schema col for entity selection

ciaranschutte opened the pull request:
**Description of changes**

- submit other entity exception types
- take schema from submitted file tsv column

**Type of Change**

- [ ] Bug
- [ ] Refactor
- [ ] New Feature
- [ ] Release Candidate

**Checklist before requesting review:**

- [ ] Check branch (code change PRs go to `develop` not master)
- [ ] Check copyrights for new files
- [ ] Manual testing
- [ ] Regression tests completed and passing (double check number of tests).
- [ ] Spelling has been checked.
- [ ] Updated swagger docs accordingly (check it's still valid)
- [ ] Set `validationDependency` in meta tag for [Argo Dictionary](https://github.com/icgc-argo/argo-dictionary) fields used in code

ciaranschutte commented on src/exception/repo/entity.ts:
```diff
@@ -45,17 +56,17 @@ const EntityExceptionModel = mongoose.model<EntityException>(
     L.debug(`Creating new donor exception with: ${JSON.stringify(exception)}`);
 
-    const update = { $set: { programId: exception.programId, specimen: exception.specimen } };
+    const update = { $set: exception };
```
upserts, so if "followUp" is saved it will not overwrite "specimen"

headsoup commented on src/exception/exception-api.ts:
```diff
@@ -62,6 +63,13 @@ const validateEntityExceptionRecords: ValidateRecords<EntityExceptionRecord> = r
+ * map submitted text from tsv schema column, to camelCase used in exception services
+ * @param record
+ * @returns camelCased schema name
+ */
```
💪I dig this new stuff you're doing, dude

headsoup reviewed (dismissed):
didn't test it, but looking good!

joneubank commented on src/exception/exception-api.ts:
```diff
@@ -62,6 +63,13 @@ const validateEntityExceptionRecords: ValidateRecords<EntityExceptionRecord> = r
+ * @param record
+ * @returns camelCased schema name
+ */
+const getSchemaEntity = (record: EntityExceptionRecord) => EntityValues[_.camelCase(record.schema)];
```
So this is one of the places where TS let's us down. When you get the value out of `EntityValues` TS says that we are guaranteed to get one of the values from the const `EntityValues`... however `_.camelCase(record.schema)` could potentially be a value not available as a property of `EntityValues` (since `record.schema` is just a string`. So the actual return type has to be declares with `| undefined` otherwise we let errors leak in.

This is entirely a consequence of getting a value using `object[key]` type lookup. :(

joneubank commented on src/exception/exception-api.ts:
```diff
@@ -108,10 +116,17 @@ class ExceptionController {
+     * schema is valid
+     * map tsv schema string to exception schema string
+     */
+    const schemaEntity = getSchemaEntity(entityExceptionRecords[0]);
```
Because of the previous comment, this could be `undefined`

joneubank reviewed (changes requested):
Going to continue reading but wanted to give you the feedback from my review of exception-api right away.

joneubank commented on src/exception/types.ts:
```diff
@@ -137,3 +141,6 @@ export const isReadonlyArrayOf = <T>(
   return input.every(validator);
 };
+
+// utility
```
This gets used frequently, it deserves a better explanation for any TS newby that interacts with it.

joneubank commented on src/exception/exception-service.ts:
```diff
@@ -29,29 +29,50 @@ import {
-  entity: Entity;
 }) => {
-  const exception: EntityException = { programId, specimen: [], followup: [] };
+  const exception: OnlyRequired<EntityException, 'programId'> = { programId };
```
Good utility type usage.

joneubank commented on src/exception/repo/entity.ts:
```diff
@@ -45,17 +56,17 @@ const EntityExceptionModel = mongoose.model<EntityException>(
 
 export interface EntityExceptionRepository {
-  save(exception: EntityException): RepoResponse<EntityException>;
+  save(exception: OnlyRequired<EntityException, 'programId'>): RepoResponse<EntityException>;
```
I really like this type utility trick.

joneubank reviewed (commented):
The rest of this was good. If we can clean up that previous issue then this is good to go.

ciaranschutte commented:
@joneubank 
I've removed the entire `getSchemaEntity` function.
The `schema` value could be any `string` until after we validate the data here: https://github.com/icgc-argo/argo-clinical/blob/feat%2F859_submit_followup_exception/src/exception/exception-service.ts#L141
So it's not a trustworthy value.

The reason to get the schema early was to include it in an error message.
Changed the error message to just be generic base case
The error handling in general needs some cleanup before the exception feature gets merged to develop.
I would like to tackle that in a "error handling ticket"
